### DISCUSSION TITLE
Possible to dynamically set paramters via args?

### DISCUSSION THREAD

**Discussion by nathanshelly** [storybookjs/storybook#13925] [Help]
👋 First off thanks for this amazing product! Migrated from v5 to v6 recently and controls inference from Typescript prop types is absolutely amazing.

I'm wondering if it's possible to set parameters of a story based on the args so that someone could play with parameters via controls.

Here's a trivial example of what I'm interested in:

```typescript
export function Foo({ fullscreen }) {
  Foo.parameters = { layout: fullscreen ? 'fullscreen' : 'centered' };

  return <div>This is a {fullscreen ? 'fullscreen' : 'centered'} component</div>;
}
Foo.args = { fullscreen: true };
```

Unfortunately setting that `parameters` inside the function doesn't have any effect in Storybook. If I add a `console.log(Foo)` before the `return <div>...` I can see that the parameters on the `Foo` object are indeed being set but Storybook isn't updating based on it.

My real use case here is with the [storybook-addon-apollo-client](https://github.com/lifeiscontent/storybook-addon-apollo-client). I would love to have a control that I could modify to change the mocks passed to the `MockedProvider` which are set on the parameters. Is there any way to achieve what I'm looking for?

**Comment by shilman** [comment]
Parameters are meant to be static, so changing them at runtime doesn't make sense.

Args are parameters that are meant to be changed at runtime. They are available everywhere parameters are available:
- As a field of the story context object `context.parameters` vs `context.args`
- Via hooks: `useParameters` vs `useArgs`

So I think rather than using args to set parameters, you can just use args instead. Would that work for your use case?

**Reply by nathanshelly** [reply]
I believe that wouldn't work for my use case as the addon I'm trying to use appears to only expose its API via parameters which your response appears to suggest are impossible to modify at runtime. Given that response seems like my use case might need changes on the addon side or I may just need to take a different tack. Appreciate your response!

**Reply by shilman** [reply]
Interesting @nathanshelly. I'd say args weren't written for this use case. `storybook-addon-apollo-client` could be modified to consume them potentially, though I don't think it's really consistent with most uses of args. Another approach is to use `globals`, which is a bit more standard as a way of communicating dynamic data to addons. But is restricted to global data, such as "theme selection" or "language selection" (see https://storybook.js.org/docs/react/essentials/toolbars-and-globals#globals). @lifeiscontent @tmeasday any thoughts here?

**Reply by tmeasday** [reply]
This is similar to a discussion @ghengeveld and I had recently with regard to his psuedo-state addon. He used globals for a reasonably similar purpose:

1.  A value that the user wants to change (in his case via a toolbar)
2. That controls the rendering of a story via a decorator
3. But that *is not* a prop of the underlying component.

Because of 1-2, an arg seems appropriate, but the problem is that if addons start defining args, then user have to start doing stuff like this in templates/stories (to avoid prop type warnings):

```js
const Template = ({ apollo, psuedoState, ...args }) => <Component {...args} />
```

*FOR EVERY COMPONENT*.

So, it seems like globals are a better solution, even if possibly you need *reset* the global every time a new story is selected (which isn't really the way globals are intended to be used).

I'm not sure if we need to give more thought to these use cases that fall between args and globals.

Anyways, all these are considerations for how you/someone would need to change the addon @nathanshelly. I'm not sure there's much you can do if the addon just uses parameters.

**Reply by tmeasday** [reply]
Maybe the solution is indeed to make parameters mutable. But that might just muddy the waters even further.

**Reply by lifeiscontent** [reply]
@shilman I'm in the same boat (args are static), for me personally, I'd just create another story.

**Reply by jamesonhill** [reply]
Has anyone figured out a hack or workaround to get dynamic/mutable parameters? Before I found this thread, I logged this issue: https://github.com/storybookjs/storybook/issues/15325. The Args announcement & docs make it seem like whatever you can do with knobs you can also accomplish using Args. Maybe it was a bug with the old version, but using knobs to control parameters has been hugely valuable but not possible with Args it seems.

**Reply by tmeasday** [reply]
@jamesonhill I replied to your issue: https://github.com/storybookjs/storybook/issues/15325#issuecomment-866455168

TLDR here is that the parameter was not changing, the parameter was a function that needed some way to read the **current value** of the arg. I proposed a way to do it for the use case presented (a GraphQL resolver mock).

**Reply by jamesonhill** [reply]
To anyone else following this, the comment here worked perfectly for me: https://github.com/storybookjs/storybook/issues/15325#issuecomment-866455168
